### PR TITLE
feat(ext): Add choco extension with param support

### LINKED ISSUE #2709: feat(ext): Add choco extension with param support
**Is this a request for help?**: Adds a Windows extension example; in this case an extension which installs packages passed as parameters via the Choco package manager. Very similar to how the one that works in DevTestLabs.

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

Links to P.R. https://github.com/Azure/acs-engine/pull/2707

---

**What version of acs-engine?**: v0.15.2

---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)** Kubernetes

**What happened**: Installs specified choco package.

**What you expected to happen**: Installs specified choco package.

**How to reproduce it** (as minimally and precisely as possible): Run cluster definiton then az group deployment.

**Anything else we need to know**: Confirmed and tested with microsoft-build-tools and it provisioned by Windows with the package.

I tested this on a generated cluster definition on top of v0.15.2

I did have to fix an entry in azuredeploy.json to get everything to work so think still need a fix in code.

```sh
"name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')), '/cse')]",
```

Should be this instead:

```sh
 "name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')),'/cse', copyIndex(variables('windowspool1Offset')))]",
```

### PR DESCRIPTION
**Pull request by sylus**
**What this PR does / why we need it**: Adds a Windows extension example; in this case an extension which installs packages passed as parameters via the Choco package manager. Very similar to how the one that works in DevTestLabs.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2709 

**Special notes for your reviewer**: Confirmed and tested with microsoft-build-tools and it provisioned by Windows with the package.

I tested this on a generated cluster definition on top of v0.15.2

I did have to fix an entry in azuredeploy.json to get everything to work so think still need a fix in code.

```sh
"name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')), '/cse')]",
```

Should be this instead:

```sh
 "name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')),'/cse', copyIndex(variables('windowspool1Offset')))]",
```

**If applicable**:
- [x] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
@sylr is this PR ready for review? Or does this mean that the code is not ready:
> I did have to fix an entry in azuredeploy.json to get everything to work so think still need a fix in code.

?

**Comment by sylus** [comment]
@CecileRobertMichon there is still the one manual adjustment that needs to be made, i can try to fix this myself but was hoping someone could point me to where that line is being generated?

Thanks so much!

**Comment by sylus** [comment]
Hey @CecileRobertMichon 

If the artifact has only the following line:

```sh
"name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')), '/cse')]",
```

Then the choco installation won't install saying can't find cse extension. This is why I need to change the line to:

```sh
 "name": "[concat(variables('windowspool1VMNamePrefix'), copyIndex(variables('windowspool1Offset')),'/cse', copyIndex(variables('windowspool1Offset')))]",
```

Thanks for pointing me to the line, i'll send an updated P.R. shortly! Appreciate your guidance as always!

**Comment by CecileRobertMichon** [comment]
@sylus can please share the exact error message you're seeing? It's curious that the choco extension relies on the CSE

**Comment by sylus** [comment]
@CecileRobertMichon here you go ^_^

```sh
❯ az group deployment create --name "acs-default" \
                           --resource-group "acs-default" \
                           --template-file "./_output/k8s-acs-default/azuredeploy.json" \
                           --parameters "./_output/k8s-acs-default/azuredeploy.parameters.json"
Deployment template validation failed: 'The resource 'Microsoft.Compute/virtualMachines/42410k8s9010/extensions/cse0' is not defined in the template. Please see https://aka.ms/arm-template for usage details.'.
```

With the mentioned fix everything then deploys correctly with choco installing the specified package.

Here was the base cluster def I used:

```sh
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.9",
      "orchestratorVersion": "1.9.6",
      "kubernetesConfig": {
        "kubernetesImageBase": "k8s-gcrio.azureedge.net/",
        "addons": [
          {
            "name": "tiller",
            "enabled" : true
          },
          {
            "name": "kubernetes-dashboard",
            "enabled" : true
          },
          {
            "name": "rescheduler",
            "enabled" : true
          }
        ]
      }
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "k8s-acs-default",
      "vmSize": "Standard_D4s_v3",
      "OSDiskSizeGB": 200
    },
    "agentPoolProfiles": [
      {
        "name": "linuxpool1",
        "count": 1,
        "customNodeLabels": {
          "os": "linux"
        },
        "vmSize": "Standard_D4s_v3",
        "OSDiskSizeGB": 200,
        "storageProfile" : "ManagedDisks",
        "availabilityProfile": "AvailabilitySet"
      },
      {
        "name": "windowspool1",
        "count": 1,
        "osType": "Windows",
        "customNodeLabels": {
          "os": "windows"
        },
        "vmSize": "Standard_D4s_v3",
        "OSDiskSizeGB": 200,
        "storageProfile" : "ManagedDisks",
        "availabilityProfile": "AvailabilitySet",
        "extensions": [
          {
            "name": "choco"
          }
        ]
      }
    ],
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "ssh-rsa XXXX"
          }
        ]
      }
    },
    "extensionProfiles": [
      {
        "name": "choco",
        "version": "v1",
        "extensionParameters": "microsoft-build-tools",
        "rootURL": "https://raw.githubusercontent.com/govcloud/acs-engine/extension-choco/"
      }
    ],
    "windowsProfile": {
      "adminUsername": "azureuser",
      "adminPassword": "XXXX",
      "imageVersion": "2016.127.20180220",
      "WindowsImageSourceUrl": ""
    },
    "servicePrincipalProfile": {
      "clientId": "XXXX",
      "secret": "XXXX"
    }
  }
}
```

File: extensions/choco/v1/template-link.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+	"name": "[concat(EXTENSION_TARGET_VM_NAME_PREFIX, copyIndex(EXTENSION_LOOP_OFFSET), 'Choco')]",
+	"type": "Microsoft.Resources/deployments",
+	"apiVersion": "[variables('apiVersionLinkDefault')]",
+	"dependsOn": [
+		"[concat('Microsoft.Compute/virtualMachines/', EXTENSION_TARGET_VM_NAME_PREFIX, copyIndex(EXTENSION_LOOP_OFFSET), '/extensions/cse', copyIndex(EXTENSION_LOOP_OFFSET))]"
```
Comment (CecileRobertMichon) [inline extensions/choco/v1/template-link.json]: This is what is causing your error, you need to make sure that the name of the VM extension is the same as it is defined in `parts\k8s\kuberneteswinagentresourcesvmas.t`

File: extensions/choco/v1/template.json [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		}
+   },
+   "variables": {  },
+   "resources": [
+	{
+	  "apiVersion": "[parameters('apiVersionCompute')]",
+	  "dependsOn": [],
+	  "location": "[resourceGroup().location]",
+	  "type": "Microsoft.Compute/virtualMachines/extensions",
+	  "name": "[concat(parameters('targetVMName'),'/cse', parameters('vmIndex'))]",
```
Comment (CecileRobertMichon) [inline extensions/choco/v1/template.json]: This should be named something more specific to chocolatey not to be confused with the already existing CSE

**Comment by CecileRobertMichon** [comment]
@sylus what is the status of this PR?

**Comment by sylus** [comment]
Ah thanks for reminding me, I will take a look at this over the weekend and have it rdy to merge given for fix before the start of Monday! @CecileRobertMichon

**Comment by CecileRobertMichon** [comment]
Deployed a test cluster w/ the extension at https://jenkins.azure-containers.io/view/acs-engine%20ad-hoc/job/k8s-deployment/102 

@JiangtianLi could you kindly review this when you have time?

**Pull request merged**